genq: add nemesis queue as an alternate shmem queue

hzhou opened the pull request:
## Pull Request Description

We add an alternate shmem queue implementation to genq so that we can easily config and test performances. The new queue is named `nem_queue` as it is essentially the exact same implementation as the queue used in ch3 nemesis.

There is only one commit that adds the `nem_queue`. The rest are refactoring and preparations to make adding or removing queue implementations more maintainable.

Addition: last 4 commits implements `free_queue` to replace the original polling-method of allocating free cells.

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->
[skip warnings]
## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
Measuring shmem bandwidth with 1024-byte messages on EPYC:

pre-3199: 2849 +/- 55 MB/s
pre-this-PR: 2341 +/- 93 MB/s
post-this-PR: 2743 +/- 50 MB/s

On Skylake:
pre-3199: 1096 +/- 44 MB/s
pre-PR: 1561 +/- 18 MB/s
post-PR: 1732 +/- 62 MB/s

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -99,16 +99,9 @@ static inline int MPIDU_genq_nem_enqueue(MPIDU_genq_shmem_queue_t queue, void *c
 
 /* INVERSE QUEUE */
 
-static inline void *MPIDU_genq_shmem_queue_head(MPIDU_genq_shmem_queue_t queue);
```
We need to keep these interfaces. SHM collectives is going to need these.

hzhou replied:
With MPSC queue, the queue head access needs thread safety. So having this API, in addition to needed implementation for each queue (does it even require implementation from each such as serial queue), it will require enqueue code be placed inside the same critical section. 

Can we add this API at the time that we are going to use it? Then we can properly review and balance the complexity and performance concerns.

A separate question: why do we need serial queue? What is the planned use case?

hzhou replied:
Pending @yfguo

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -52,6 +52,30 @@ static inline MPIDU_genqi_shmem_cell_header_s
     for (void *tmp = MPIDU_genq_shmem_queue_head((queue)), cell = tmp; tmp; \
          tmp = MPIDU_genq_shmem_queue_next(tmp))
 
+static inline int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue,
```
Why is queue init moved this moved to `MPIDU_genq_shmem_pool.c`?

hzhou replied:
invalid.

yfguo commented on src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.c:
```diff
@@ -64,8 +64,8 @@ int MPIDI_POSIX_iqueue_init(int rank, int size)
-    mpi_errno = MPIDU_genq_shmem_queue_init(transport->my_terminal, transport->cell_pool,
-                                            MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC);
+    mpi_errno = MPIDU_genq_shmem_queue_init(transport->my_terminal,
+                                            MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM);
```
Naming of queue type should be general. Not referring to NEM. In fact, I think it is better to just delete inversed queue. Then this can simply be `MPSC` queue type.

hzhou replied:
done.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_pool.h:
```diff
@@ -13,14 +13,10 @@
 #include <stdint.h>
 #include <string.h>
 
-typedef void *MPIDU_genq_shmem_pool_t;
```
This transparent pool handle type should be in `mpidu_genq_shmem_pool.h`. The user should be not be looking at the internal `mpidu_genqi_shmem_types.h` for it.

hzhou replied:
User are not accessing the fields of `shmem_pool` nor `shmem_queue`. Both of them are opaque to user. I guess we are arguing for compiler enforcement, which I think it is a minor issue and not worth extra code just for the sake of it.

yfguo replied:
The meaning of this is that `MPIDU_genq_shmem_pool_t pool` becomes an input parameter semantically.

hzhou replied:
I see. Will fix.

hzhou replied:
Both pool and queue have been fully abstracted. Queue no longer depends on pool.

yfguo commented on src/mpid/common/genq/mpidu_genqi_shmem_types.h:
```diff
@@ -9,13 +9,43 @@
+    } q;
+    uint8_t pad[3 * MPIDU_SHM_CACHE_LINE_LEN];
+} MPIDU_genq_shmem_queue_u;
+
```
The queue union should be exposed in the public `mpidu_genq_shmem_queue.h`. Otherwise, the user would have not idea how much memory is needed for the queue object.

hzhou replied:
User doesn't need know, the compiler knows :) , and that is sufficient.

yfguo replied:
The user need to know this because it is user's job to allocate the memory for the queue obj. If we only exposing the transparent queue handle type `MPIDU_genq_shmem_queue_t`, the user won't know what is size of the queue obj.

hzhou replied:
I see. So why don't we have `MPIDU_genq_shmem_queue_alloc` (which probably can combine with `MPIDU_genq_shmem_queue_init`) ?

yfguo replied:
The user may want to allocated queues on shared memory collective (e.g. pt2pt recv queue use case) or they may want to allocated the queues independently (e.g. SHM collective use case).

If queue is tied to the shmem pool, it would be forcing the collective allocation model.

hzhou replied:
I didn't mean it to be tied to the shmem pool. The shmem queue has to be allocated collectively on shared memory anyway, so instead of user call `MPIDU_Init_shm_alloc`, which not `MPIDU_genq_shmem_queue_init` call it for the user in stead?

hzhou replied:
> If queue is tied to the shmem pool, it would be forcing the collective allocation model.

Oh, just realized that `pool` is already tied to the queue. It is a parameter to `enqueue` `dequeue` since the base address on different processors are different.

yfguo replied:
The queue object need to be on shared memory. The shared memory need to be collectively allocated. If we do a queue allocation function, that function (assuming it allocates shmem) to be collective.

By only initializing a queue obj on an allocated shared memory region is more flexible in that we can init a new queue object without other process participating.

yfguo replied:
The collective allocation of queue might makes sense for recv queues. But it is to strict for SHM collectives queues.

hzhou replied:
> 
> 
> The collective allocation of queue might makes sense for recv queues. But it is too strict for SHM collectives queues.

Pre-mature optimization is the root of all evil. How about let's not worry about collective shmem until that is being worked at? The worst case, we always can create new interface, e.g `MPIDU_genq_shemem_queue_get_size()`, or we can rework current API.

yfguo replied:
We shouldn't do that because we will be removing it. Think about this: if process 0 is going to allocated a shmem queue while process 1 is not going to create a shmem queue, why should we force process 1 to participate the queue allocation? We could simply let it init from a preallocated slab.

yfguo replied:
For example, SHM bcast would be the root enqueue the data into to its own queue, notify everyone to read the cell enqueued. Only the root process needs to init the queue. The other process should not participating the creation of that queue.

hzhou replied:
> why should we force process 1 to participate the queue allocation?

That is a straw man argument. When we have that situation, we can work at the solution then. Let's not to base our design on something that may or may not be important.

hzhou replied:
> 
> 
> For example, SHM bcast would be the root enqueue the data into to its own queue, notify everyone to read the cell enqueued. Only the root process needs to init the queue. The other process should not participating the creation of that queue.

We can't init the queue at init time?

yfguo replied:
Unfortunately no, there is going to be per-comm.

hzhou replied:
If you want to talk about collective shmem, we can discuss offline. I don't think collective shmem as you hinted has any merit, and definitely we shouldn't use some very vague future code to make current argument.

yfguo commented on src/mpid/common/genq/mpidu_genqi_shmem_types.h:
```diff
@@ -9,13 +9,43 @@
 #include "mpidu_init_shm.h"
 
+typedef void *MPIDU_genq_shmem_pool_t;
+typedef void *MPIDU_genq_shmem_queue_t;
```
The transparent queue handle should be exposed in the public mpidu_genq_shmem_queue.h.

hzhou replied:
See above.

hzhou commented:
@yfguo I am addressing your comments aggregated here.

* Removing inverse queue (or inverted queue as Intel refers to it)
I somehow believe Intel has studied the performance when they introduced iqueue. Since they are the stakeholder simply removing it in a non-plugable way may be disruptive. I would like at least some feedbacks before removing it.

* Splitting init code into .c files has negligible benefits (saving compile time?) and it will result in spread of code related to queue algorithms in a interleaved fashion. Since the queue implementations uses similar fields with subtle nuisances, having them spread and interleaved is too much for the negligible benefits. On the other hand, having them grouped together makes reading much easier and  adding/removing queue algorithm much simpler -- if you consider deleting or adding a single block of code as one change.

* Naming
    ** The enum names can be changed if you like. They serve for readability and distinction. Meeting such requirements, I would prefer shorter for more readability. `MPSC` does not serve distinction, and the usage is always clear on what type of queue they needed, so I personally don't think it needs more suffix.

   **  Now that the actual queue functions (`_init`, `_enqueue`, `_dequeue`) are internal functions, the naming just need serve the readability and distinction within this file. They will follow what the ENUM names we decide. 

So for the naming, propose your set of names and I'll update them throughout.

yfguo commented:
> @yfguo I am addressing your comments aggregated here.
> 
> * Removing inverse queue (or inverted queue as Intel refers to it)
>   I somehow believe Intel has studied the performance when they introduced iqueue. Since they are the stakeholder simply removing it in a non-plugable way may be disruptive. I would like at least some feedbacks before removing it.

Agree.

> * Splitting init code into .c files has negligible benefits (saving compile time?) and it will result in spread of code related to queue algorithms in a interleaved fashion. Since the queue implementations uses similar fields with subtle nuisances, having them spread and interleaved is too much for the negligible benefits. On the other hand, having them grouped together makes reading much easier and  adding/removing queue algorithm much simpler -- if you consider deleting or adding a single block of code as one change.

I would prefer the `MPIDU_genq_shmem_queue_init` in `mpidu_genq_shmem_queue.c`. It is no functionality or naming correlation with `pool`.

> * Naming
>   ** The enum names can be changed if you like. They serve for readability and distinction. Meeting such requirements, I would prefer shorter for more readability. `MPSC` does not serve distinction, and the usage is always clear on what type of queue they needed, so I personally don't think it needs more suffix.
>   **  Now that the actual queue functions (`_init`, `_enqueue`, `_dequeue`) are internal functions, the naming just need serve the readability and distinction within this file. They will follow what the ENUM names we decide.
>   So for the naming, propose your set of names and I'll update them throughout.

MPSC is an appropriate name. The inverted queue is designed for MPSC. Also it has the benefit of pointing out that the queue is for MPSC. I suggest do the following name changes:

`MPIDU_GENQ_SHMEM_QUEUE_TYPE__INVERSE` => `MPIDU_GENQ_SHMEM_QUEUE_TYPE__INV_MPSC`
`MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM` => `MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC`

`MPIDU_genq_serial_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_serial_init,dequeue,enqueue`
`MPIDU_genq_nem_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_mpsc_init,dequeue,enqueue`
`MPIDU_genq_inverse_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_inv_mpsc_init,dequeue,enqueue`

hzhou commented:
> I would prefer the `MPIDU_genq_shmem_queue_init` in `mpidu_genq_shmem_queue.c`. It is no functionality or naming correlation with `pool`.

Why do you refer to `pool`? I just consolidated  `mpidu_genq_shmem_queue.c` into `mpidu_genq_shmem_queue.h`.

> MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM => MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC

Don't you find `MPSC` too generic? What if later we want to introduce a variation of `MPSC` queue? or even a variation of nemesis queue?

yfguo commented:
> > I would prefer the `MPIDU_genq_shmem_queue_init` in `mpidu_genq_shmem_queue.c`. It is no functionality or naming correlation with `pool`.
> 
> Why do you refer to `pool`? I just consolidated `mpidu_genq_shmem_queue.c` into `mpidu_genq_shmem_queue.h`.

I miss read the line. Sorry.

> 
> > MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM => MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC
> 
> Don't you find `MPSC` too generic? What if later we want to introduce a variation of `MPSC` queue? or even a variation of nemesis queue?

So far we only have (REGULAR)_MPSC and INV_MPSC. If we add a variant of MPSC, it is going to me SOMEVARIANT_MPSC.  It is easier by looking at the name and know that it is MPSC queue. Names like NEM or INVERSE are missing the information about MPSC which is crucial.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -206,4 +238,76 @@ static inline void *MPIDU_genq_shmem_queue_next(void *cell)
 
+/* EXTERNAL INTERFACE */
+
+static inline int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue,
```
`mpidu_genq_shmem_queue.c` now is kept.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -206,4 +238,76 @@ static inline void *MPIDU_genq_shmem_queue_next(void *cell)
+
+    if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
+        rc = MPIDU_genq_serial_init(queue);
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC) {
```
done.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -206,4 +238,76 @@ static inline void *MPIDU_genq_shmem_queue_next(void *cell)
+        rc = MPIDU_genq_serial_init(queue);
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC) {
+        rc = MPIDU_genq_inverse_init(queue);
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM) {
```
done.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -52,8 +84,8 @@ static inline MPIDU_genqi_shmem_cell_header_s
 
-static inline int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue,
-                                              MPIDU_genq_shmem_pool_t pool, int flags)
+static inline int MPIDU_genq_inverse_init(MPIDU_genq_shmem_queue_t queue,
```
outdated.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -276,7 +277,7 @@ static inline int MPIDU_genq_shmem_queue_dequeue(MPIDU_genq_shmem_queue_t queue,
     if (queue->q.flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
         rc = MPIDU_genq_serial_dequeue(queue, cell);
-    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC) {
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__INVERSE) {
```
done.

hzhou commented:
@yfguo I have updated the PR, addressing *most* of your comments. Also I rearranged the commits so that refactoring commits (without functional changes) are consolidated and grouped together, which may make the PR easier to review.

Could you review it again?

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch4/ucx

yfguo commented:
> > @yfguo I am addressing your comments aggregated here.
> > 
> > * Removing inverse queue (or inverted queue as Intel refers to it)
> >   I somehow believe Intel has studied the performance when they introduced iqueue. Since they are the stakeholder simply removing it in a non-plugable way may be disruptive. I would like at least some feedbacks before removing it.
> 
> Agree.
> 
> > * Splitting init code into .c files has negligible benefits (saving compile time?) and it will result in spread of code related to queue algorithms in a interleaved fashion. Since the queue implementations uses similar fields with subtle nuisances, having them spread and interleaved is too much for the negligible benefits. On the other hand, having them grouped together makes reading much easier and  adding/removing queue algorithm much simpler -- if you consider deleting or adding a single block of code as one change.
> 
> I would prefer the `MPIDU_genq_shmem_queue_init` in `mpidu_genq_shmem_queue.c`. It is no functionality or naming correlation with `pool`.
> 
> > * Naming
> >   ** The enum names can be changed if you like. They serve for readability and distinction. Meeting such requirements, I would prefer shorter for more readability. `MPSC` does not serve distinction, and the usage is always clear on what type of queue they needed, so I personally don't think it needs more suffix.
> >   **  Now that the actual queue functions (`_init`, `_enqueue`, `_dequeue`) are internal functions, the naming just need serve the readability and distinction within this file. They will follow what the ENUM names we decide.
> >   So for the naming, propose your set of names and I'll update them throughout.
> 
> MPSC is an appropriate name. The inverted queue is designed for MPSC. Also it has the benefit of pointing out that the queue is for MPSC. I suggest do the following name changes:
> 
> `MPIDU_GENQ_SHMEM_QUEUE_TYPE__INVERSE` => `MPIDU_GENQ_SHMEM_QUEUE_TYPE__INV_MPSC`
> `MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM` => `MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC`
> 
> `MPIDU_genq_serial_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_serial_init,dequeue,enqueue`
> `MPIDU_genq_nem_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_mpsc_init,dequeue,enqueue`
> `MPIDU_genq_inverse_init,dequeue,enqueue` => `MPIDU_genqi_shmem_queue_inv_mpsc_init,dequeue,enqueue`

@hzhou It looks like `MPIDU_genq_serial_init,dequeue,enqueue` is renamed to ``MPIDU_genqi_serial_init,dequeue,enqueue` in the latest revision. Can you add "shmem" to the name?

hzhou commented:
> @hzhou It looks like `MPIDU_genq_serial_init,dequeue,enqueue` is renamed to ``MPIDU_genqi_serial_init,dequeue,enqueue` in the latest revision. Can you add "shmem" to the name?

These are internal functions that are only defined and used inside this file. Rather than making the names longer, I was thinking of making the names even shorter :) e.g. `genqi_serial_init`, etc. They only need be distinct with each other, and `genqi` is sufficient as namespace to prevent collision, the `shmem` or `queue` is very clear from the context of `mpidu_genq_shmem_queue.h`, so there is no benefit of adding it other than noise.

hzhou commented:
@yfguo Added last four commits to address better abstractions:

* genq: add shmem queues alloc/free functions
  -- make queue struct fully internal and opaque 
* genq: rename mpidu_genqi_shmem_types.h and clean up
  -- internal header doesn't work with inline functions. Use proper markups in stead.
* genq: rename internal symbols with genqi prefix
  -- Better names to clearly distinguish external/internal struct and functions
* genq: better queues abastraction
  -- make queues opaque and independent of pools

hzhou commented:
test:mpich/ch4/ofi

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.c:
```diff
@@ -4,28 +4,48 @@
 
-int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue, MPIDU_genq_shmem_pool_t pool,
-                                int flags)
+int MPIDU_genq_shmem_queues_alloc(int num_of_queues, MPIDU_genq_shmem_queues_t * queues)
```
I don't think this function is necessary. It wraps shmem allocation and make it a shmem calloc with fixed element size. 

Can we just expose the `sizeof(MPIDU_genq_shmem_queue_u)` as a macro, so the user can handle shmem allocation without explicitly knowing the `MPIDU_genq_shmem_queue_u` type.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.c:
```diff
@@ -4,28 +4,48 @@
+}
+
+/* initialize queue on a existing shared memeory */
+int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queues_t queues, int queue_idx, int flags)
```
If we remove the "queues_alloc", we can revert this to simply init queue on a given address.

hzhou replied:
I am removing this instead -- (making it internal at least). I am not convinced that we ever need break the abstraction and let user init a bare queue on a piece of shared memory.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.c:
```diff
@@ -4,28 +4,48 @@
+    goto fn_exit;
+}
 
+int MPIDU_genq_shmem_queues_free(MPIDU_genq_shmem_queues_t queues)
```
Would not need if alloc is removed.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_types.h:
```diff
@@ -0,0 +1,99 @@
+    ((genqi_shmem_cell_header_s *) ((char *) (cell) - sizeof(genqi_shmem_cell_header_s)))
+
+/* The handle value is calculated using (address_offset + 1), see MPIDU_genq_shmem_pool.h */
+#define HANDLE_TO_HEADER(base, handle) \
```
I don't see what this change is needed.

hzhou replied:
Detach the dependency of queue from pool. I thought you argued for it.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -83,103 +155,92 @@ static inline MPIDU_genqi_shmem_cell_header_s
-    if (cell_h) {
-        return HEADER_TO_CELL(cell_h);
+static inline int MPIDU_genq_shmem_queue_dequeue(MPIDU_genq_shmem_queues_t queues,
+                                                 int queue_idx, void **cell)
```
This interface change is strange. It is not a general queue interface. It seems have a specific usecase in mind.

yfguo replied:
This queues + idx interface is quite counter-intuitive to general queue.

hzhou replied:
Otherwise, you need expose some internal details. Shmem queue is unique from normal queues, I think one can get used to it.

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -83,103 +155,92 @@ static inline MPIDU_genqi_shmem_cell_header_s
+    if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL) {
+        rc = genqi_serial_dequeue(queues.base, queue_obj, cell);
+    } else if (flags == MPIDU_GENQ_SHMEM_QUEUE_TYPE__INV_MPSC) {
+        rc = genqi_inv_mpsc_dequeue(queues.base, queue_obj, cell);
```
`queues.base` was not initialized in `queues_alloc`.

hzhou replied:
Wait for my last commit...

yfguo commented on src/mpid/common/genq/mpidu_genq_shmem_types.h:
```diff
@@ -0,0 +1,99 @@
+typedef struct MPIDU_genq_shmem_queues_t {
+    void *base;                 /* base address for the cells */
+    void *ptr;                  /* pointer to shmem allocation */
+} MPIDU_genq_shmem_queues_t;
```
Essentially, this is creating a local array of reference to the shared queue object. The purpose seems to be avoiding passing pool as a parameter for enqueue/dequeue operation?

hzhou replied:
You don't need pool for enqueue/dequeue, you only need base address.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.c:
```diff
@@ -8,13 +8,37 @@
 #include "mpidu_genqi_shmem_types.h"
 
-int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queue_t queue, int flags)
+int MPIDU_genq_shmem_queues_alloc(int num_of_queues, MPIDU_genq_shmem_queue_t * queues_out)
```
MACRO doesn't abstract the internal data type.

I am pushing another commit to further abstract `MPIDU_genq_shmem_queue_t` to `void *`, hang on...

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_types.h:
```diff
@@ -79,7 +83,7 @@ typedef struct genqi_shmem_pool {
     void *slab;
     genqi_shmem_cell_header_s *cell_header_base;
```
`MPIDU_genq_` and `MPIDU_genqi_` are just too similar and takes too much effort to distinguish. I find it much easier to read by just use `genqi_` to mark internal. As we agreed, internal symbols has no use of `MPIDU_` prefix.

hzhou commented on src/mpid/common/genq/mpidu_genq_shmem_queue.h:
```diff
@@ -7,56 +7,128 @@
+
+/* SERIAL */
+
+static inline int genqi_serial_init(genqi_shmem_queue_u * queue)
```
Please don't spread the algorithm apart. One has to read the init and enqueue/dequeue together to understand the algorithm.

hzhou commented:
@yfguo We probably can be more effective resolving a few key issues offline:
* shmem collectives
* how to deal with pure internal symbols within inline headers
* what should be the proper abstraction

yfguo commented:
> @yfguo We probably can be more effective resolving a few key issues offline:
> * shmem collectives
> * how to deal with pure internal symbols within inline headers
> * what should be the proper abstraction

Sure. We can do a call on Monday.

hzhou commented:
test:mpich/ch4/most

hzhou commented:
We (@yfguo) had an offline discussion. Much of the disagreement is about how the collective shmem will work, which will only use the serial queue interface. The serial queue and mpsc queue doesn't share much commonality and cannot be interchanged, so it appears if we simply split them into, e.g. `MPIDU_shmem_queue_serial.h` and `MPIDU_shmem_queue_mpsc.h`, then most of the infrastructure could be removed. Much of the past confusion and debate wouldn't happen in the first place. So I believe we should do that as an follow up PR.

The point being raised that `genq` will no longer provide any "generality". I think if an interface doesn't bring benefits, then there is no point keep it. I think we can simply use `mpid/common/genq` as a source code organization for pool and queue related code but host inside a collection of standalone utility implementations. Having standalone utilities helps the code orthogonality already.

hzhou commented:
@wesbland This PR swaps the inverted queue algorithm. Would you or your team like to take a review?

hzhou commented:
> I didn't review any of the genq refactoring work here. I was just looking at the other changes, which I think are fine, but I didn't do any performance checks yet. Have you?

Yes, we did. See https://github.com/pmodels/mpich/pull/4753#issuecomment-678827098

wesbland commented:
> Yes, we did. See [#4753 (comment)](https://github.com/pmodels/mpich/pull/4753#issuecomment-678827098)

That's OSU BW?

hzhou commented:
> > Yes, we did. See [#4753 (comment)](https://github.com/pmodels/mpich/pull/4753#issuecomment-678827098)
> 
> That's OSU BW?

Adapted from OSU BW, focusing on message size of 1024 byte. The OSU BW curves behave similar.

raffenet reviewed (approved):
Looks good. Thanks for staying on top of this!
